# Tutorial on an example dataset

In this tutorial, we provide step-by-step instructions to run the benchmarking pipeline. The necessary pod5 and references are provided in the example folder:<br/> 
- Pod5 files - [example/pod5](example/pod5) <br/>
- Reference - [example/references/H.pylori_J99_ATCC700824.fa](example/references/H.pylori_J99_ATCC700824.fa)

You can also run the pipeline using your own dataset by replacing the example dataset folder with your dataset folder and downloading the appropriate genome file from UCSC.

Creating conda environment

## Conda environment setup and deployment

```bash

# Activate conda environment
conda create \
    -n ont_basemod_benchmarking_env \
    -c conda-forge \
    -c bioconda \
     snakemake

# Activate conda environment

conda activate ont_basemod_benchmarking_env
```

## Basic dependencies required to run the workflow
- snakemake
- [samtools](https://www.htslib.org/) >=v1.16
- [bedtools](https://bedtools.readthedocs.io/en/latest/) 
- [pod5-api](https://pod5-file-format.readthedocs.io/en/0.1.21/docs/install.html)

<b>Note: </b>all of these can be installed directly using the environment setup yml as follows:
```bash
    conda env create -f env.yml -n ont_basemod_benchmarking_env
    conda activate ont_basemod_benchmarking_env
```

## Install tools to be benchmarked using installation instructions for each tool. 

GitHub links for methylation calling tools are given below:

- [Dorado](https://github.com/nanoporetech/dorado)<br/>
- [DeepMod2](https://github.com/WGLab/DeepMod2/tree/main)<br/>
- [DeepBAM](https://github.com/xiaochuanle/DeepBAM)<br/>
- [DeepPlant](https://github.com/xiaochuanle/DeepPlant/tree/main)<br/>
- [f5C](https://github.com/hasindu2008/f5c/tree/master)<br/>
- [Rockfish](https://github.com/lbcb-sci/rockfish/tree/r10.4.1)<br/>


## Running snakemake workflow

Before executing the snakemake pipeline, add appropriate paths to the tools and the datasets in the config.yaml file. 
The snakemake file contains all the rules for the snakemake pipeline to run. Run the following commad to create the output files:

The following is the directory structure generated by the snakemake workflow. Of which, those highlighted cyan are required by the workflow to start running while the others are generated by the workflow as it progresses through the pipeline.

<pre>
.
├── <span style="color: cyan">workflow</span>        <span style="color: green"># snakemake workflow directory </span>
├── <span style="color: cyan">example</span>         <span style="color: green"># Example folder </span>
├── <span style="color: cyan">pod5</span>            <span style="color: green"># Default path to pod5 files</span>
├── bam             <span style="color: green"># basecalled mod/move_table bams</span>
├── modkit          <span style="color: green"># modkit output</span>
├── deepmod2        <span style="color: green"># deepmod2 output</span>
├── deepbam         <span style="color: green"># deepbam output</span>
├── deepplant       <span style="color: green"># deepplant output</span>
├── f5c             <span style="color: green"># f5c output</span>
├── rockfish        <span style="color: green"># rockfish output</span>
├── log             <span style="color: green"># log folder generated by workflow</span>
├── meta            <span style="color: green"># standardised bed output</span>
├── meta_context    <span style="color: green"># standardised bed output with C context</span>
├── qc              <span style="color: green"># qc folder</span>
├── <span style="color: cyan">references.yaml</span> <span style="color: green"># list of references to use</span>
└── <span style="color: cyan">config.yaml</span>     <span style="color: green"># snakemake config file</span>
</pre>

### Running the example

```bash

    conda activate ont_basemod_benchmarking_env
    cd example

    snakemake --use-conda -s ../snakemake_pipelines/snakefile --cores 10
```
**Note:** For more detailed usage of the workflow check the [snakemake.md] (snakemake_pipelines/snakemake.md) file